docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,319 @@
+
+#### Insecure providers
+- An insecure provider MUST not implement certificate rotation or kubelet serving certificate signing
+- InfraCluster.Status.KubeletClientCertificateSigner MUST be set to cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure
+- An insecure provider MUST use the cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure signer/v1alpha4
+
+#### Secure providers
+- A secure provider MUST implement certificate rotation and kubelet server certificate signing
+- A provider must register signers of:
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-<provider>/v1alpha4`
+    - `cluster-x-k8s-io/kubelet-serving-<provider>/v1alpha4`
+- A secure provider MUST implement a secure attestation mechanism, based upon PEM-encoded blocks within the Certificate Signing Request.
+- Where a secure provider’s attestation mechanism does not include a nonce or timestamp to protect against replay attacks, the mechanism MUST implement a secondary time-limited attestation (e.g. AWS Instance Identity document + AWS HMACv4 signature).
+- A provider’s signer MUST run on the management cluster
+
+#### TPM based providers
+- A TPM provider MUST use the following certificate signers
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-tpm/v1alpha4`
+    - `cluster-x-k8s-io/kubelet-serving-tpm/v1alpha4`
+- A TPM provider MUST annotate new CSRs as follows:
+    - Key: cluster-x-k8s-io/tpm-endorsement-key/v1alpha4
+    - Value: Platform-specific endorsement key (e.g., retrieved from GCP Shielded VM API or VMware vCenter)
+
+#### Kubeadm
+Since this proposal essentially takes over part of the node registration process from kubeadm, we will require the following changes:
+- kubeadm COULD allow opt-out of kubeadm setting up ClusterRoleBindings between the system:nodes group and the `system:certificates.k8s.io:certificatesigningrequests:selfnodeclient` permission, so that certificate renewals must go through re-attestation. However, Cluster API kubeadm control plane provider is able to reconcile the removal of these roles during control plane instantiation.
+- Kubeadm COULD allow opt-out of kubeadm setting up the bootstrapper and node groups and bindings.
+
+
+#### Kubelet authenticator flow
+
+The authenticator will be responsible for both updating the kubelet client and serving certificates.
+
+![kubelet auth](images/kubelet-authentication/kubelet-auth.png)
+
+
+
+### User Stories
+
+
+
+#### Story 1: machine attestation
+
+A cluster operator has been asked to ensure compliance with NIST SP 800-190 Application Container Security Guide. Hardware countermeasure 4.6 suggests that container platforms should make use of trusted computing. In a Kubernetes context, this would mean providing hardware node attestation wherever possible. Each cloud provider supports different mechanisms for this (see Appendix). (cluster-api/3762
```

Comment:
a link to the NIST PDF would be nice here.

Reply:
the trailing `(cluster-api/3762` seems incomplete.

Reply:
Done , cleaned up.